[Go] Implement GetProofByHash

gdbelvin opened the pull request:
This PR includes the code required to fetch inclusion proofs

Contributes to #1159

gdbelvin commented:
Thoughts on the best way to test this are welcome.

gdbelvin commented:
Added tests

pphaneuf commented:
Another merge crept up, these muddle up the PRs a bit, you should use something along the lines of `git fetch https://github.com/google/certificate-transparency.git master && git rebase FETCH_BASE` (could be simplified a bit, but the exact command-line would depend on what the remotes on your local repo are named).

gdbelvin commented:
Rebased as requested. PTAL

gdbelvin commented:
Rebased again. PTAL 
@pphaneuf

pphaneuf commented on go/client/logclient.go:
```diff
@@ -105,6 +101,19 @@ type getEntryAndProofResponse struct {
 
+type getProofByHashResponse struct {
+	LeafIndex    int      `json:"leaf_index"` // The 0-based index of the end entity corresponding to the "hash" parameter.
+	AuditPath    []string `json:"audit_path"` // An array of base64-encoded Merkle Tree nodes proving the inclusion of the chosen certificate.
```
Does it work if you make this field a `[][]byte` instead (and strip out the base64 decoding at line 423)?

pphaneuf commented on go/client/logclient.go:
```diff
@@ -105,6 +101,19 @@ type getEntryAndProofResponse struct {
+	ErrorMessage string   `json:"error_message"`
+}
+
+// addJSONRequest represents the JSON request body sent ot the add-json CT
```
Why did this move?

pphaneuf commented on go/client/logclient.go:
```diff
@@ -105,6 +101,19 @@ type getEntryAndProofResponse struct {
 }
 
+type getProofByHashResponse struct {
+	LeafIndex    int      `json:"leaf_index"` // The 0-based index of the end entity corresponding to the "hash" parameter.
```
We're using `int64` for leaf indices in general, is there a problem with the JSON parser that prevents using that?

pphaneuf commented on go/client/logclient.go:
```diff
@@ -386,3 +403,28 @@ func (c *LogClient) GetEntries(start, end int64) ([]ct.LogEntry, error) {
+	if err != nil {
+		return nil, err
+	}
+	if resp.ErrorMessage != "" {
```
I think you might be disappointed by some recent developments in `fetchAndParse` (and might need a further rebase, at the minimum)...

pphaneuf commented on go/serialization.go:
```diff
@@ -126,10 +130,10 @@ func readASN1CertList(r io.Reader, totalLenBytes int, elementLenBytes int) ([]AS
 // Returns a non-nil error if there was a problem.
 func ReadTimestampedEntryInto(r io.Reader, t *TimestampedEntry) error {
 	var err error
```
That line might not be needed anymore, with your other changes? (I like your way better!)

pphaneuf commented on go/serialization.go:
```diff
@@ -144,13 +148,48 @@ func ReadTimestampedEntryInto(r io.Reader, t *TimestampedEntry) error {
 			return err
 		}
+	case XJSONLogEntryType:
+		if t.JSONData, err = readVarBytes(r, JSONLengthBytes); err != nil {
```
But if you do that, might have to make the `=` into a `:=` as well?

gdbelvin replied:
Kept the original scheme.  Has to do with `struct.field` not being something you can `:=` assign.

pphaneuf commented:
Hmm... Am I missing some reference from `logclient.go`, or are the changes in `serialization.go`, `types.go`, and `signatures.go` are not related at all to what's in the description of this PR? Other change got mixed in?

gdbelvin commented:
The changes in `types.go` and `serialization.go` allow a client to compute what the right hash of a given `SCT` and `json` so that the client can use the `GetByHash` function.   I decided to separate the compute hash and get-by-hash functions in case the client has the hash available directly. 

Cleaned up, rebased, updated. PTAL.

gdbelvin commented:
PTAL

gdbelvin commented:
Fixes #1278

eranmes commented on go/client/logclient.go:
```diff
@@ -85,13 +82,21 @@ type getAcceptedRootsResponse struct {
+type GetProofByHashResponse struct {
+	LeafIndex    int64    `json:"leaf_index"` // The 0-based index of the end entity corresponding to the "hash" parameter.
+	AuditPath    [][]byte `json:"audit_path"` // An array of base64-encoded Merkle Tree nodes proving the inclusion of the chosen certificate.
+	Success      bool     `json:"success"`
```
I suggest getting rid of this and the ErrorMessage because the GetProofByHashResponse should not be created if there was a failure.

eranmes commented on go/merkletree/merkle_verifier_test.go:
```diff
@@ -299,9 +310,30 @@ func getConsistencyProofs() []consistencyTestVector {
+	// {"leaf_index":0,"audit_path":[]}
+
+	// Compute leaf value from timestamp
+	if err := ct.SerializeV1SCTMerkleTreeLeafX509(googleLeaf, uint64(1469742100562), googlePEM); err != nil {
```
I believe you could use test data to load certificate, SCT from here:
https://github.com/google/certificate-transparency/tree/master/test/testdata
The test will then be much simpler and will only test that SerializeV1SCTMerkleTreeLeafX509 correctly serializes.

gdbelvin commented:
@pphaneuf @taknira 

Cleaned up this PR, and made it exclusively about fetching the inclusion proof.

pphaneuf commented on go/types.go:
```diff
@@ -319,6 +319,7 @@ type TimestampedEntry struct {
 	X509Entry    ASN1Cert
 	JSONData     []byte
 	PrecertEntry PreCert
+	JSONData     []byte
```
Same as the one at line 320? This causes a build error. :wink:

gdbelvin commented:
Fixed

pphaneuf commented on go/client/logclient.go:
```diff
@@ -308,3 +317,12 @@ func (c *LogClient) GetSTH() (sth *ct.SignedTreeHead, err error) {
 }
+
+// GetProofByHash returns an audit path for the hash of an SCT.
+func (c *LogClient) GetProofByHash(hash []byte, treeSize uint64) (GetProofByHashResponse, error) {
```
Big enough that `GetProofByHashResponse` should be returned as a pointer?

pphaneuf replied:
That would also allow returning `nil` in the case of an error.

pphaneuf replied:
We have some historical baggage, but might as well take a `context.Context` as the first parameter on new methods?

pphaneuf commented on go/client/logclient.go:
```diff
@@ -308,3 +317,12 @@ func (c *LogClient) GetSTH() (sth *ct.SignedTreeHead, err error) {
+
+// GetProofByHash returns an audit path for the hash of an SCT.
+func (c *LogClient) GetProofByHash(hash []byte, treeSize uint64) (GetProofByHashResponse, error) {
+	var resp GetProofByHashResponse
```
I'd move that to right before it's needed (before `fetchAndParse`).

pphaneuf commented on go/types.go:
```diff
@@ -317,8 +317,8 @@ type TimestampedEntry struct {
 	X509Entry    ASN1Cert
-	JSONData     []byte
 	PrecertEntry PreCert
+	JSONData     []byte
```
Maybe you could move it up to where it was, to avoid the noise?

pphaneuf commented:
Done with this round of review, should be straightforward... :smiley:

pphaneuf commented on go/client/logclient.go:
```diff
@@ -319,10 +319,12 @@ func (c *LogClient) GetSTH() (sth *ct.SignedTreeHead, err error) {
-func (c *LogClient) GetProofByHash(hash []byte, treeSize uint64) (GetProofByHashResponse, error) {
-	var resp GetProofByHashResponse
+func (c *LogClient) GetProofByHash(ctx context.Context, hash []byte, treeSize uint64) (*GetProofByHashResponse, error) {
+	resp := new(GetProofByHashResponse)
```
Oops, you've seen my comment in #1285? I like `var` better than `new`: it seems there's no case where `new` does anything that you can't do with `var` and `&`, so I'd rather have just one way of doing things, for consistency, and let the compiler do the escape analysis...

And either way, I'd still want it declared "at the last moment". :wink:

pphaneuf commented on go/client/logclient_test.go:
```diff
@@ -275,7 +275,7 @@ func TestGetProofByHash(t *testing.T) {
 
 	for _, tc := range tests {
-		resp, err := c.GetProofByHash(tc.hash, tc.treesize)
+		resp, err := c.GetProofByHash(context.TODO(), tc.hash, tc.treesize)
```
I guess that in a test, it should be `context.Background()`? Not like we'll refactor in the future to get the context from somewhere else, right?

pphaneuf commented on go/serialization.go:
```diff
@@ -147,6 +147,10 @@ func ReadTimestampedEntryInto(r io.Reader, t *TimestampedEntry) error {
 		if t.PrecertEntry.TBSCertificate, err = readVarBytes(r, PreCertificateLengthBytes); err != nil {
 			return err
 		}
+	case XJSONLogEntryType:
```
I kind of missed this in the previous round, but, er, this has nothing to do with "implementing GetProofByHash", doesn't it? :wink:

gdbelvin replied:
Would you like it moved to a different PR? It was so nice to sneak it into this one !

pphaneuf replied:
Yeah, this is making this PR all icky! :laughing:

pphaneuf commented:
LGTM, waiting on Travis to merge.

pphaneuf commented:
Will need a rebase (and you can squash into one commit, if you want).

gdbelvin commented:
Rebased on master
